crates/runtime/src/iam_auth.rs

```diff
@@ -0,0 +1,1389 @@
+async fn generate_gcp_tokens(config: &GCPConfig, task_name: &str) -> anyhow::Result<GCPTokens> {
+    // Google presents the audience with https:, so we strip that if it exists
+    let aud = config
+        .gcp_workload_identity_pool_audience
+        .strip_prefix("https:")
+        .unwrap_or(&config.gcp_workload_identity_pool_audience);
+
+    // Step 1: Sign a JWT using the default runtime service account with task_name in payload
+    let mut signed_jwt = google_sign_jwt(task_name, None, aud).await?;
+
+    // Step 2: Exchange the signed JWT for an access token via OAuth 2.0 token exchange
+    let mut exchanged_token = exchange_jwt_for_service_account_token(&signed_jwt, aud).await?;
+
+    signed_jwt.zeroize();
+
+    // Step 3: Use the exchanged access token to impersonate the target service account
+    let impersonated_token =
+        impersonate_service_account(&exchanged_token, &config.gcp_service_account_to_impersonate)
+            .await?;
+
+    exchanged_token.zeroize();
+
+    Ok(GCPTokens {
+        access_token: impersonated_token,
+    })
+}
+
+/// Get GCP access token from service account credentials JSON
+async fn get_gcp_token_from_credentials(credentials_json: &str) -> anyhow::Result<String> {
+    use serde_json::Value;
+
+    let key_data: Value = serde_json::from_str(credentials_json)
+        .context("failed to parse service account key JSON")?;
+
+    let client_email = key_data
+        .get("client_email")
+        .and_then(|v| v.as_str())
+        .context("missing client_email in service account key")?;
+
+    let private_key = key_data
+        .get("private_key")
+        .and_then(|v| v.as_str())
+        .context("missing private_key in service account key")?;
+
+    create_service_account_jwt_token(client_email, private_key).await
+}
+
+/// Sign a JWT using Google's signJWT API with configurable subject and audience
+async fn google_sign_jwt(
+    task_name: &str,
+    subject: Option<&str>,
+    audience: &str,
+) -> anyhow::Result<String> {
+    let credentials_path = std::env::var("GOOGLE_APPLICATION_CREDENTIALS")
+        .context("GOOGLE_APPLICATION_CREDENTIALS environment variable not set")?;
+
+    if credentials_path.is_empty() {
+        anyhow::bail!("GOOGLE_APPLICATION_CREDENTIALS environment variable is empty");
+    }
+
+    let mut credentials_json = tokio::fs::read_to_string(&credentials_path)
+        .await
+        .with_context(|| {
+            format!(
+                "Failed to read Google Cloud credentials from {}",
+                credentials_path
+            )
+        })?;
+
+    if credentials_json.trim().is_empty() {
+        anyhow::bail!("Google Cloud credentials file is empty");
+    }
+
+    // Parse the credentials to get the runtime service account email
+    let key_data: serde_json::Value = serde_json::from_str(&credentials_json)
+        .context("failed to parse service account key JSON")?;
+    let runtime_service_account = key_data
```

Comment:
using `service_account_email` as the name might quiet the auto code scanning alerts? :shrug: